### PR TITLE
ONS Design System feature toggle

### PR DESCRIPTION
**Pull request by rav-pradhan**
### What

Added a new feature flag, `ONSDesignSystemVersion`, has been added to the base `Page` model. This value can be set to a specific version of the ONS design system (as a string). If the string is not empty, `dp-renderer` will take the value and load the CSS and JS for the ONS design system in tandem to Sixteens.

This is applied on a per-page basis via the page's mapper function. This allows for new pages that use the design system components to be built accordingly, whilst minimising issues across the ONS website that arise due to style conflicts between Sixteens and the design system.

### How to review

- Sense check template and page model changes

#### To test locally

- Pull this branch down and checkout
- Open `dp-frontend-cookie-controller` or `dp-frontend-dataset-controller` (or any other controller that has since been migrated to use `dp-renderer`)
- Use the `replace` keyword in `go.mod` to use your local version of `dp-renderer` ([guide here if you're unfamiliar with this](https://thewebivore.com/using-replace-in-go-mod-to-point-to-your-local-module/))
- Run the relevant frontend app with `make debug`
- Visit a page and inspect the network. You should see no reference to the ONS design system CSS and JS files being fetched from the CDN.
- Update the mapper function of that page to include the line `p.FeatureFlags.ONSDesignSystemVersion = "37.0.0"`
- Reload the app, refresh the page and inspect the network - you should see the ONS design system CSS and JS files fetched from the CDN

### Who can review

Anyone

### CODE REVIEW

File: assets/templates/main.tmpl [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+  <head>
 
-    <title>{{ if .Metadata.Title -}}{{ .Metadata.Title }}
-      {{ else if .Error -}}{{if .Error.Title }}{{ .Error.Title }}{{- end }}
+    <title>
+      {{ if .Metadata.Title -}}{{ .Metadata.Title }}
+      {{ else if .Error -}}
+        {{if .Error.Title }}{{ .Error.Title }}
+        {{- end }}
       {{- end}} - {{ localise "OfficeForNationalStatistics" .Language 1 }}</title>
```
Comment (mr-nick17) [inline assets/templates/main.tmpl:10]: Put the close tag on a new line, it starts as a block level element

File: assets/templates/main.tmpl [hunk 2]
```diff
@@ -22,33 +32,62 @@
...
-    <script type="text/javascript" src="/js/app.js"></script>
+  {{ if .FeatureFlags.ONSDesignSystemVersion }}
+      </div>
+    </div>
+    <script>
+      var ONS_assets_base_URL = 'https://cdn.ons.gov.uk/sdc/design-system/{{.FeatureFlags.ONSDesignSystemVersion}}/';
+    </script>
+    <script src="https://cdn.ons.gov.uk/sdc/design-system/{{.FeatureFlags.ONSDesignSystemVersion}}/scripts/main.js"></script>
+  {{end}}
+  <script type="text/javascript" src="{{.PatternLibraryAssetsPath}}/js/main.js"></script>
```
Comment (mr-nick17) [inline assets/templates/main.tmpl:90]: `type` attribute is not needed for scripts

File: assets/templates/main.tmpl [hunk 3]
```diff
@@ -22,33 +32,62 @@
...
+  <script type="text/javascript" src="/js/app.js"></script>
```
Comment (mr-nick17) [inline assets/templates/main.tmpl:91]: `type` attribute is not needed for scripts

File: assets/templates/main.tmpl [hunk 4]
```diff
@@ -22,33 +32,62 @@
...
 </body>
 </html>
```
Comment (mr-nick17) [inline assets/templates/main.tmpl:93]: No empty line at the end of the file

File: model/page.go [hunk 5]
```diff
@@ -25,7 +25,8 @@ type Page struct {
 }
```
Comment (mr-nick17) [inline model/page.go:26]: Public types should have a godoc

**Comment by mr-nick17** [comment]
The design system also introduces a console error, related to cookies. As cookies are handled by a separate service, I would suggest that the cookie control JS files should be excluded. 

Screenshot 2021-08-10 at 11 03 34

**Comment by rav-pradhan** [comment]
> The design system also introduces a console error, related to cookies. As cookies are handled by a separate service, I would suggest that the cookie control JS files should be excluded.
> 
> Screenshot 2021-08-10 at 11 03 34

Yeah this is a good point. I think namespacing will be the quickest and least intrusive way to resolve this e.g., `cookies-banner` is the class used to hook the JS interaction to the element; if we updated in Sixteens to be `dp-cookies-banner` or the design system was updated to be, for instance, `ons-cookies-banner`, then our own logic would override the logic in the ds

**Review by mr-nick17 — APPROVED** [review]
LGTM

**Pull request merged**